File: scripts/migrate-voice-data.py
#!/usr/bin/env python3
"""
Migration script to import voice time data from old bot
Reads from data/persistence_data.json and populates voice_time_aggregates table

ONE-TIME USE SCRIPT - Hardcoded guild ID
"""

import json
import os
import psycopg2
from psycopg2.extras import execute_values

# Hardcoded guild ID for migration
GUILD_ID = '367904135548239872'

def parse_time_to_seconds(time_string):
    """
    Parses time string format "days:hours:minutes:seconds" to total seconds
    Example: "76:04:59:37" = (76 days * 86400) + (4 hours * 3600) + (59 min * 60) + 37 sec
    """
    parts = [int(x) for x in time_string.split(':')]

    if len(parts) != 4:
        raise ValueError(f"Invalid time format: {time_string}")

    days, hours, minutes, seconds = parts
    return (days * 86400) + (hours * 3600) + (minutes * 60) + seconds

def extract_user_id_from_voice_key(key):
    """
    Extracts user_id from key like "223647480741363713_voice"
    Returns None if not a voice key
    """
    if not key.endswith('_voice'):
        return None

    return key.replace('_voice', '')

def migrate_voice_data():
    """Main migration function"""
    print('📊 Starting voice time data migration...\n')

    # Read and parse JSON file
    json_path = './data/persistence_data.json'
    with open(json_path, 'r') as f:
        data = json.load(f)

    print(f'🏰 Guild ID: {GUILD_ID}\n')

    # Extract voice data
    voice_data = {}

    # Process lifetime_sums
    for key, time_string in data['lifetime_sums'].items():
        user_id = extract_user_id_from_voice_key(key)
        if user_id:
            lifetime_seconds = parse_time_to_seconds(time_string)
            voice_data[user_id] = {
                'lifetime_seconds': lifetime_seconds,
                'weekly_seconds': 0
            }

    # Process this_week_time_sums
    for key, time_string in data['this_week_time_sums'].items():
        user_id = extract_user_id_from_voice_key(key)
        if user_id:
            weekly_seconds = parse_time_to_seconds(time_string)
            if user_id in voice_data:
                voice_data[user_id]['weekly_seconds'] = weekly_seconds
            else:
                # User has weekly data but no lifetime data (shouldn't happen, but handle it)
                voice_data[user_id] = {
                    'lifetime_seconds': weekly_seconds,
                    'weekly_seconds': weekly_seconds
                }

    print(f'👥 Found {len(voice_data)} users with voice time data\n')

    # Hardcoded database credentials (REPLACE WITH YOUR VALUES)
    conn = psycopg2.connect(
        host='YOUR_AWS_RDS_HOST.rds.amazonaws.com',
        port='5432',
        database='hogbot',
        user='hogbot',
        password='YOUR_DATABASE_PASSWORD'
    )

    try:
        cur = conn.cursor()

        # Ensure guild exists in database
        cur.execute(
            """INSERT INTO guild_settings (guild_id)
               VALUES (%s)
               ON CONFLICT (guild_id) DO NOTHING""",
            (GUILD_ID,)
        )

        # Prepare data for batch insert
        insert_data = []
        for user_id, times in voice_data.items():
            insert_data.append((
                user_id,
                GUILD_ID,
                times['lifetime_seconds'],
                times['weekly_seconds']
            ))

        # Batch insert with ON CONFLICT
        insert_query = """
            INSERT INTO voice_time_aggregates (user_id, guild_id, total_seconds, weekly_seconds, weekly_updated_at)
            VALUES %s
            ON CONFLICT (user_id, guild_id) DO UPDATE
            SET total_seconds = EXCLUDED.total_seconds,
                weekly_seconds = EXCLUDED.weekly_seconds,
                weekly_updated_at = NOW(),
                updated_at = NOW()
        """

        execute_values(
            cur, insert_query, insert_data,
            template="(%s, %s, %s, %s, NOW())"
        )

        conn.commit()

        print(f'\n✅ Migration complete!')
        print(f'   📥 Inserted/Updated: {len(insert_data)} users')

        # Show sample of migrated data
        print('\n📊 Sample migrated data:')
        cur.execute(
            """SELECT user_id, total_seconds, weekly_seconds
               FROM voice_time_aggregates
               WHERE guild_id = %s
               ORDER BY total_seconds DESC
               LIMIT 5""",
            (GUILD_ID,)
        )

        print('\nTop 5 users by total voice time:')
        for row in cur.fetchall():
            user_id, total_seconds, weekly_seconds = row
            hours = total_seconds // 3600
            weekly_hours = weekly_seconds // 3600
            print(f'  User {user_id}: {hours:,} hours total, {weekly_hours:,} hours this week')

        cur.close()

    except Exception as e:
        conn.rollback()
        print(f'❌ Migration failed: {e}')
        raise
    finally:
        conn.close()

if __name__ == '__main__':
    try:
        migrate_voice_data()
    except Exception as e:
        print(f'Fatal error: {e}')
        exit(1)

File: scripts/migrate-voice-data.ts
/**
 * Migration script to import voice time data from old bot
 * Reads from data/persistence_data.json and populates voice_time_aggregates table
 *
 * ONE-TIME USE SCRIPT - Hardcoded guild ID
 */

import { readFileSync } from 'fs';
import { pool } from '../src/lib/database.js';

interface PersistenceData {
  lifetime_sums: Record<string, string>;
  this_week_time_sums: Record<string, string>;
}

// Hardcoded guild ID for migration
const GUILD_ID = '367904135548239872';

/**
 * Parses time string format "days:hours:minutes:seconds" to total seconds
 * Example: "76:04:59:37" = (76 days * 86400) + (4 hours * 3600) + (59 min * 60) + 37 sec
 */
function parseTimeToSeconds(timeString: string): number {
  const parts = timeString.split(':').map(Number);

  if (parts.length !== 4) {
    throw new Error(`Invalid time format: ${timeString}`);
  }

  const [days, hours, minutes, seconds] = parts;

  return (days * 86400) + (hours * 3600) + (minutes * 60) + seconds;
}

/**
 * Extracts user_id from key like "223647480741363713_voice"
 * Returns null if not a voice key
 */
function extractUserIdFromVoiceKey(key: string): string | null {
  if (!key.endsWith('_voice')) {
    return null;
  }

  return key.replace('_voice', '');
}

async function migrateVoiceData() {
  const client = await pool.connect();

  try {
    console.log('📊 Starting voice time data migration...\n');

    // Read and parse JSON file
    const jsonPath = './data/persistence_data.json';
    const rawData = readFileSync(jsonPath, 'utf-8');
    const data: PersistenceData = JSON.parse(rawData);

    console.log(`🏰 Guild ID: ${GUILD_ID}\n`);

    // Ensure guild exists in database
    await client.query(
      `INSERT INTO guild_settings (guild_id)
       VALUES ($1)
       ON CONFLICT (guild_id) DO NOTHING`,
      [GUILD_ID]
    );

    // Extract voice data
    const voiceData = new Map<string, { lifetimeSeconds: number; weeklySeconds: number }>();

    // Process lifetime_sums
    for (const [key, timeString] of Object.entries(data.lifetime_sums)) {
      const userId = extractUserIdFromVoiceKey(key);
      if (userId) {
        const lifetimeSeconds = parseTimeToSeconds(timeString);
        voiceData.set(userId, { lifetimeSeconds, weeklySeconds: 0 });
      }
    }

    // Process this_week_time_sums
    for (const [key, timeString] of Object.entries(data.this_week_time_sums)) {
      const userId = extractUserIdFromVoiceKey(key);
      if (userId) {
        const weeklySeconds = parseTimeToSeconds(timeString);
        const existing = voiceData.get(userId);

        if (existing) {
          existing.weeklySeconds = weeklySeconds;
        } else {
          // User has weekly data but no lifetime data (shouldn't happen, but handle it)
          voiceData.set(userId, { lifetimeSeconds: weeklySeconds, weeklySeconds });
        }
      }
    }

    console.log(`👥 Found ${voiceData.size} users with voice time data\n`);

    // Insert data into database
    await client.query('BEGIN');

    let insertedCount = 0;
    let skippedCount = 0;

    for (const [userId, { lifetimeSeconds, weeklySeconds }] of voiceData.entries()) {
      try {
        await client.query(
          `INSERT INTO voice_time_aggregates (user_id, guild_id, total_seconds, weekly_seconds, weekly_updated_at)
           VALUES ($1, $2, $3, $4, NOW())
           ON CONFLICT (user_id, guild_id) DO UPDATE
           SET total_seconds = EXCLUDED.total_seconds,
               weekly_seconds = EXCLUDED.weekly_seconds,
               weekly_updated_at = NOW(),
               updated_at = NOW()`,
          [userId, GUILD_ID, lifetimeSeconds, weeklySeconds]
        );

        insertedCount++;

        // Log progress every 10 users
        if (insertedCount % 10 === 0) {
          console.log(`✅ Migrated ${insertedCount} users...`);
        }
      } catch (error) {
        console.error(`❌ Failed to insert user ${userId}:`, error);
        skippedCount++;
      }
    }

    await client.query('COMMIT');

    console.log('\n✅ Migration complete!');
    console.log(`   📥 Inserted/Updated: ${insertedCount} users`);
    if (skippedCount > 0) {
      console.log(`   ⚠️  Skipped: ${skippedCount} users`);
    }

    // Show sample of migrated data
    console.log('\n📊 Sample migrated data:');
    const sample = await client.query(
      `SELECT user_id, total_seconds, weekly_seconds
       FROM voice_time_aggregates
       WHERE guild_id = $1
       ORDER BY total_seconds DESC
       LIMIT 5`,
      [GUILD_ID]
    );

    console.log('\nTop 5 users by total voice time:');
    for (const row of sample.rows) {
      const hours = Math.floor(row.total_seconds / 3600);
      const weeklyHours = Math.floor(row.weekly_seconds / 3600);
      console.log(`  User ${row.user_id}: ${hours.toLocaleString()} hours total, ${weeklyHours.toLocaleString()} hours this week`);
    }

  } catch (error) {
    await client.query('ROLLBACK');
    console.error('❌ Migration failed:', error);
    throw error;
  } finally {
    client.release();
    await pool.end();
  }
}

// Run migration
migrateVoiceData().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
